src/mlpack/methods/ann/layer/multihead_attention.hpp

```diff
@@ -0,0 +1,321 @@
+                arma::Mat<eT>& gradient);
+
+  /**
+   * Serialize the layer
+   */
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned int /* version */);
+
+  //! Get the embedding dimension.
+  size_t EmbedDim() const { return embedDim; }
+  //! Modify the embedding dimension.
+  size_t& EmbedDim() { return embedDim; }
+
+  //! Get the number of attention heads.
+  size_t NumHeads() const { return numHeads; }
+  //! Modify the number of attention heads.
+  size_t& NumHeads() { return numHeads; }
+
+  //! Get the key matrix.
+  InputDataType const& Key() const { return key; }
+  //! Modify the key matrix.
+  InputDataType& Key() { return key; }
+
+  //! Get the value matrix.
+  InputDataType const& Value() const { return value; }
+  //! Modify the value matrix.
+  InputDataType& Value() { return value; }
+
+  //! Get the two dimensional Attention Mask.
+  InputDataType const& AttentionMask() const { return attnMask; }
+  //! Modify the two dimensional Attention Mask.
+  InputDataType& AttentionMask() { return attnMask; }
+
+  //! Get Key Padding Mask.
+  InputDataType const& KeyPaddingMask() const { return keyPaddingMask; }
+  //! Modify the Key Padding Mask.
+  InputDataType& KeyPaddingMask() { return keyPaddingMask; }
+
+  //! Get the output parameter.
+  OutputDataType const& OutputParameter() const { return outputParameter; }
+  //! Modify the output parameter.
+  OutputDataType& OutputParameter() { return outputParameter; }
+
+  //! Get the delta.
+  OutputDataType const& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  //! Get the gradient.
+  OutputDataType const& Gradient() const { return grad; }
+  //! Modify the gradient.
+  OutputDataType& Gradient() { return grad; }
+
+  //! Get the parameters.
+  OutputDataType const& Parameters() const { return weights; }
+  //! Modify the parameters.
+  OutputDataType& Parameters() { return weights; }
+
+
+ private:
+  //! Element Type of the input.
+  typedef typename InputDataType::elem_type ElemType;
+
+  template <typename eT>
+  void Expand(const arma::Mat<eT>& query,
+              arma::Mat<eT>& key,
+              arma::Mat<eT>& value,
+              arma::Cube<eT>& q,
+              arma::Cube<eT>& k,
+              arma::Cube<eT>& v)
+  {
+    typedef typename arma::Cube<eT> CubeType;
+
+    Log::Assert(query.n_rows % embedDim == 0, "Number of features in 'query' \
+        must be divisible by embedding dimension");
+    Log::Assert(key.n_rows % embedDim == 0, "Number of features in 'key' \
+        must be divisible by embedding dimension");
+    Log::Assert(arma::size(key) == arma::size(value), "'key' and 'value' \
+        must have same dimensions.");
+
+    q = CubeType(const_cast<arma::Mat<eT>&>(query).memptr(),
+        query.n_rows / embedDim, embedDim, query.n_cols, true, false);
+
+    if (key.is_empty())
+      k = CubeType(q.memptr(), q.n_rows, q.n_cols, q.n_slices, false, false);
```

Comment:
>Why don't you extract key and value from the input like you do with query?

We can get `k` from `input` as you suggested but `v` must be extracted from `k` only. Because when `value` is empty (line 223) it doesn't mean that `key` is also empty. And in that case `value = key` i.e. the Encoder-Decoder Attention case. Let me know if I missed something.

>I thought we obtain query, key, and value by the forward propagation.

You are correct. But I think it is possible that we train the encoder part first to get the `key` and use it to train the decoder part. (At least this should work for Transformer Model).

Reply:
> We can get k from input as you suggested but v must be extracted from k only.

I think we can implement it in a layer that will be before `MultiHeadAttention`. In this case, `MultiHeadAttention` would accept input of shape `(tgtSeqLen + 2 * srcSeqLen, embedDim, batchSize)`. I mean we could introduce a layer that copies `query` to `key`, `key` to `value` depending on the attention type. I think this could simplify the backward pass in `MultiHeadAttention` since the layer I've just proposed will handle this.

> You are correct. But I think it is possible that we train the encoder part first to get the key and use it to train the decoder part.

Ok, let's leave this PR as is and leave it open until we start the implementation of the decoder and the encoder. I feel we'll need to change the API of `MultiHeadAttention`.

Reply:
>In this case, MultiHeadAttention would accept input of shape (tgtSeqLen + 2 * srcSeqLen, embedDim, batchSize).

I tried this. Using this method, none of the cases pass for the Jacobian Test. We can see why from this [comment](https://github.com/mlpack/mlpack/pull/2375#discussion_r466137134). This method was very tedious (maybe for me) to implement in mlpack. The reason being Forward function accepts just two arguments, one input, and one output.

>I mean we could introduce a layer that copies query to key, key to value depending on the attention type. I think this could simplify the backward pass in MultiHeadAttention since the layer I've just proposed will handle this.

I'm not getting this one. Can you please illustrate this?

>Ok, let's leave this PR as is

That's scary 😬. Let's get it done by this week, please 😂. Sorry for the conflicting idea but we can't do anything in Transformer model without multihead attention. It's at the core. And we are on the verge to get it done. :+1: 

>leave it open until we start the implementation of the decoder and the encoder.

I have already opened a PR in models repository https://github.com/mlpack/models/pull/16. But how will we proceed without the MultiheadAttention class?

Reply:
> I'm not getting this one. Can you please illustrate this?

I mean the positional encoding layer.

> But how will we proceed without the MultiheadAttention class?

The problem is about passing the input to `MultiHeadAttention`. How are you going to implement that?